pkg/codegen/go/gen.go

```diff
@@ -1350,8 +1350,12 @@ func (pkg *pkgContext) genObjectDefaultFunc(w io.Writer, name string,
 			if err != nil {
 				return err
 			}
-			pkg.needsUtils = true
-			fmt.Fprintf(w, "if isZero(tmp.%s) {\n", pkg.fieldName(nil, p))
+			if isNilType(p.Type) {
+				fmt.Fprintf(w, "if tmp.%s == nil {\n", pkg.fieldName(nil, p))
+			} else {
+				pkg.needsUtils = true
+				fmt.Fprintf(w, "if isZero(tmp.%s) {\n", pkg.fieldName(nil, p))
+			}
```

Comment:
Can we simplify this to remove the else case? 

Or rephrased, do we support defaults for non-optional fields?

Reply:
Grepping the PR branch, I only see 4 remaining calls to `isZero`:

* three in the `plain-and-default` test in `plain-and-default/go/foo/moduleResource.go`
* one in the `plain-object-defaults` test in `plain-object-defaults/go/example/pulumiTypes.go`

It seems safe to merge this if tests pass but I do wonder what exactly we're testing in those 4 cases. All of the fields involved are non-pointer, required fields in the schema. The codegen even looks nonsensical:

```
	if isZero(args.Plain_required_string) {
		args.Plain_required_string = "buzzer"
	}
```

I can almost see the argument for string or integer, but "isZero(args.Plain_required_boolean)" is nonsensical.

Reply:
For string and integer values, I wonder if we should instead be codegenning this for the zero case of required strings and booleans:

```
	if isZero(args.FieldName) {
		return nil, errors.New("invalid value for required argument 'ResourceNoAlias'")
	}
```

Or if we should be rejecting these programs during codegen as having invalid schemas. A required field with a default does not compute, unless the goal is to allow any value _except_ the zero value, forbidding empty strings, 0 and 0.0.

Reply:
We have existing providers that set all combinations of `required`, `const` and `default` for the same field. We are keeping `isZero` and allowing these cases for backwards compatibility reasons. Fixing this is on the list of 4.0 changes.